Lombiq.Tests.UI/Extensions/VisualVerificationUITestContextExtensions.cs

```diff
@@ -0,0 +1,567 @@
+                diff.AbsoluteError,
+                diff.MeanError,
+                diff.PixelErrorCount,
+                diff.PixelErrorPercentage),
+            messageIfExists: HintFailureDumpItemAlreadyExists);
+
+        comparator(diff);
+    }
+
+    private static void AssertInternal<TValue>(
+        VisualVerificationMatchApprovedContext approvedContext,
+        Func<TValue, TValue, bool> comparer,
+        TValue actual,
+        TValue expected,
+        string propertyName,
+        string condition)
+    {
+        if (comparer(actual, expected))
+        {
+            return;
+        }
+
+        throw new VisualVerificationAssertionException(
+            FormatAssertionMessage(
+                approvedContext,
+                actual,
+                expected,
+                propertyName,
+                condition));
+    }
+
+    private static void AssertInternal<TValue>(
+        Func<TValue, TValue, bool> comparer,
+        TValue actual,
+        TValue expected,
+        string propertyName,
+        string condition)
+    {
+        if (comparer(actual, expected))
+        {
+            return;
+        }
+
+        throw new VisualVerificationAssertionException(
+            FormatAssertionMessage(
+                approvedContext: null,
+                actual,
+                expected,
+                propertyName,
+                condition));
+    }
+
+    private static string FormatAssertionMessage<TValue>(
+        VisualVerificationMatchApprovedContext approvedContext,
+        TValue actual,
+        TValue expected,
+        string propertyName,
+        string condition)
+    {
+        var message = new StringBuilder()
+            .AppendLine()
+            .AppendLine(
+                CultureInfo.InvariantCulture,
+                $"{propertyName} should be {condition} {expected} but the calculated value is {actual}");
+
+        if (approvedContext != null)
+        {
+            string loadedFrom = null;
+            if (!string.IsNullOrEmpty(approvedContext.ReferenceImagePath))
+            {
+                loadedFrom = $"file: {approvedContext.ReferenceImagePath}";
+            }
+            else if (!string.IsNullOrEmpty(approvedContext.ReferenceResourceName))
+            {
+                loadedFrom = $"embedded resource: {approvedContext.ReferenceResourceName}";
+            }
+
+            if (!string.IsNullOrEmpty(loadedFrom))
+            {
+                message
+            .AppendLine()
+            .AppendLine("Visual verification failed since the asserted element looks different from the reference image.")
+                    .AppendLine(
+                        CultureInfo.InvariantCulture,
+                        $"The reference image was loaded from {loadedFrom}.")
+                    .AppendLine("If you want a new reference image, simply delete the existing one ")
+                    .Append("and a new one will be generated on next run.");
```

Comment:
I previously suggested to change the second line to ˙˙Append()` because I thought then the message will be in a single line. This is the correct code for that:

```suggestion
                    .Append("If you want a new reference image, simply delete the existing one ")
                    .AppendLine("and a new one will be generated on next run.");
```